### PR TITLE
[BugFix] Fix ConcurrentModificationException error on query

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue

expect retry process:
```
1. lock 
2. analyze (with origin meta data)
3. unlock
4. transformRelation (with origin meta data)
5. lock
6. copy data (unused)
7. unlock
8. optimizer work (with origin meta data)
9. check version
10. retry
11. lock 
12. copy data
13. unlock
14. analyze (with copy meta data)
15. transformRelation (with copy meta data)
16. ...
```

First query execute, doesn't use copy data to `tranfromRelation`, and will throw ConcurrentModificationException on step.8, this PR just a simple BugFix

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 6 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|0|2|00.00%|[225, 226]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/RangePartitionInfo.java [hunk 1]
```diff
@@ -72,9 +72,9 @@ public class RangePartitionInfo extends PartitionInfo {
     @SerializedName(value = "partitionColumns")
     protected List<Column> partitionColumns = Lists.newArrayList();
     // formal partition id -> partition range
-    protected Map<Long, Range<PartitionKey>> idToRange = Maps.newHashMap();
+    protected Map<Long, Range<PartitionKey>> idToRange = Maps.newConcurrentMap();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/catalog/RangePartitionInfo.java:75]: needn't concurrent?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/catalog/RangePartitionInfo.java:75]: is unnecessary to use hash-map? and will set concurrentMap when rollback editlog now

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|6|100.00%|[]|

**Review by kangkaisen — DISMISSED** [review]

**Review by packy92 — DISMISSED** [review]

**Review by liuyehcf — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 46 / 50 (92.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|8|10|80.00%|[105, 106]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|9|11|81.82%|[316, 317]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ListPartitionInfo.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ExpressionRangePartitionInfoV2.java|6|6|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**